feat(Table): Allow authors to import table

geoffreykwan opened the pull request:
## Changes

Added the ability for authors to import a table in the Table component authoring.

## Test

1. Open a project in the Authoring Tool
2. Go to a Table step
3. Go to the Advanced view for the Table component authoring
4. Click the "Import Table" button
5. Choose a csv file from your computer (that has 500 or less cells)
6. Make sure the Table component now contains the data from your csv file
7. Try importing a table with more than 500 cells, it should not allow you to do so

Closes #684

geoffreykwan committed:
feat(Table): Allow authors to import table

#684

hirokiterashima commented on src/assets/wise5/components/table/edit-table-advanced/edit-table-advanced.component.ts:
```diff
@@ -177,4 +178,130 @@ export class EditTableAdvancedComponent extends EditAdvancedComponentComponent {
+  }
+
+  importTable(fileContent: string): void {
+    const twoDimensionalStringArray = this.csvToArray(fileContent);
```
Can we use ```UtilService.CSVToArray(strData, strDelimiter)``` here instead?

hirokiterashima reviewed (changes requested):
Functionality works for small file size I tried (2KB, 12 rows, 20 cols). It increased size of project.json from 9KB to 20KB though, which may become an issue if the unit uses a lot of large table data sets. When I tried uploading 96KB file (108 rows, 24 cols), it froze the AT and the VLE preview didn't load, and the browser eventually gave me a "do you want to kill the tabs?" alert message.

Some questions:
1. Should we limit file size and/or number of rows/columns?
2. Should we store the data in a separate file (outside of project.json) and add a link to it instead? This will reduce project.json size and only fetch the file when users need it.
3. Should we show a "importing..." message?

See inline comment.

geoffreykwan committed:
feat(Table): Add importing table message

Limit the imported table to have 500 cells or less.
Call UtilService.CSVToArray().

#684

hirokiterashima commented on src/assets/wise5/components/table/edit-table-advanced/edit-table-advanced.component.ts:
```diff
@@ -177,4 +183,91 @@ export class EditTableAdvancedComponent extends EditAdvancedComponentComponent {
+    if (confirm($localize`Are you sure you want to overwrite the existing table?`)) {
+      this.showImportingTableDisplay();
+      this.setImportTableMessage($localize`Importing table...`);
+      const files = event.target.files;
```
Move this files variable declaration to right above where we need it, or don't declare at all.
```
const files = event.target.files;
reader.readAsText(files[0]);
```
or
```
reader.readAsText(event.target.files[0]);
```

hirokiterashima commented on src/assets/wise5/components/table/edit-table-advanced/edit-table-advanced.component.ts:
```diff
@@ -177,4 +183,91 @@ export class EditTableAdvancedComponent extends EditAdvancedComponentComponent {
+      const files = event.target.files;
+      const reader: any = new FileReader();
+      reader.onload = () => {
+        const fileContent = reader.result;
```
```suggestion
      const reader: FileReader = new FileReader();
      reader.onload = () => {
        const fileContent = reader.result as string;
```

hirokiterashima commented on src/assets/wise5/components/table/edit-table-advanced/edit-table-advanced.component.ts:
```diff
@@ -177,4 +183,91 @@ export class EditTableAdvancedComponent extends EditAdvancedComponentComponent {
+      };
+      reader.readAsText(files[0]);
+    } else {
+      event.target.value = null;
```
This line is in both the if-true and if-else cases: ```event.target.value = null;``` Can we move it outside and remove the else case?

hirokiterashima commented on src/assets/wise5/components/table/edit-table-advanced/edit-table-advanced.component.ts:
```diff
@@ -177,4 +183,91 @@ export class EditTableAdvancedComponent extends EditAdvancedComponentComponent {
+    for (const row of tableContent) {
+      for (const cell of row) {
+        numCells++;
+      }
```
```suggestion
      numCells += row.length;
```

hirokiterashima commented on src/assets/wise5/components/table/edit-table-advanced/edit-table-advanced.component.ts:
```diff
@@ -177,4 +183,91 @@ export class EditTableAdvancedComponent extends EditAdvancedComponentComponent {
+    this.componentChanged();
+  }
+
+  convertToTableData(stringArray: string[][]): any[][] {
+    const table = [];
+    for (const row of stringArray) {
+      const tableRow = [];
+      for (const cell of row) {
+        tableRow.push({
+          text: cell,
+          editable: true,
+          size: null
+        });
+      }
+      table.push(tableRow);
+    }
+    return table;
+  }
```
```suggestion
  convertToTableData(stringArray: string[][]): any[][] {
    return stringArray.map((row) =>
      row.map((cell) => {
        return { text: cell, editable: true, size: null };
      })
    );
  }
```

geoffreykwan replied:
I left this as a double for loop since it will be more efficient when there are more cells. The double for loop is also less complex according to CodeMetrics.

hirokiterashima reviewed (changes requested):
New functionality looks good. See additional inline comments.

geoffreykwan committed:
refactor(Table): Clean up code

#684

hirokiterashima reviewed (approved):
Re-writing the map this way
```
    return stringArray.map((row) =>
      row.map((cell) => ({ text: cell, editable: true, size: null }))
    );
```
or this way (requires defining a new Cell class):
```
    return stringArray.map((row) =>
      row.map((cell) => new Cell(cell, true, null))
    );
```
seems to result in the same complexity as the for-loop (5 in the first way, 4 in the second if you also set function return type to Cell). We should define a Cell class in any case, in a future refactoring task.

```for...of``` loop does seem to be faster for large data sets than ```map``` ([reference](https://leanylabs.com/blog/js-forEach-map-reduce-vs-for-for_of/)), but since we are limiting the number of cells to 500 in this function, I wonder if the gains would be that much noticeable. I personally find it easier to read code that is more declarative and less verbose (5 lines with map vs 11 with for...of). But this is a personal opinion and we should discuss and agree on a convention as a team.

Approving for now, but let's continue to discuss and weigh the pros and cons of each.
